<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Tissue-specific promoter marks | Tidy Ranges Tutorial</title>
  <meta name="description" content="Basic examples of computing operations on genomic ranges using the
tidy data philosophy." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Tissue-specific promoter marks | Tidy Ranges Tutorial" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Basic examples of computing operations on genomic ranges using the
tidy data philosophy." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Tissue-specific promoter marks | Tidy Ranges Tutorial" />
  
  <meta name="twitter:description" content="Basic examples of computing operations on genomic ranges using the
tidy data philosophy." />
  

<meta name="author" content="nullranges devel team" />


<meta name="date" content="2021-12-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="overlap-granges-with-plyranges.html"/>
<link rel="next" href="bootstrap-overlap.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Tidy Ranges Tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="chapter" data-level="1" data-path="join-is-an-overlap.html"><a href="join-is-an-overlap.html"><i class="fa fa-check"></i><b>1</b> Join is an overlap</a></li>
<li class="chapter" data-level="2" data-path="overlap-granges-with-plyranges.html"><a href="overlap-granges-with-plyranges.html"><i class="fa fa-check"></i><b>2</b> Overlap GRanges with plyranges</a></li>
<li class="chapter" data-level="3" data-path="tissue-specific-promoter-marks.html"><a href="tissue-specific-promoter-marks.html"><i class="fa fa-check"></i><b>3</b> Tissue-specific promoter marks</a></li>
<li class="chapter" data-level="4" data-path="bootstrap-overlap.html"><a href="bootstrap-overlap.html"><i class="fa fa-check"></i><b>4</b> Bootstrap overlap</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tidy Ranges Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="tissue-specific-promoter-marks" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Tissue-specific promoter marks</h1>
<p>Objective: determine if promoters marks (e.g.Â H3K27ac) are specific to
genes that are expressed in a tissue-specific manner.</p>
<p>We will load expression data from the GTEx project, which gives
median expression in TPM for each tissue.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" title="1"><span class="kw">library</span>(tidyr)</a>
<a class="sourceLine" id="cb45-2" title="2">file &lt;-<span class="st"> &quot;data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz&quot;</span></a>
<a class="sourceLine" id="cb45-3" title="3">gtex &lt;-<span class="st"> </span><span class="kw">read.delim</span>(file, <span class="dt">skip=</span><span class="dv">2</span>)</a></code></pre></div>
<p>We select two tissues, bladder and kidney, and convert the data from a
wide format into a tidy format.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" title="1">tissues &lt;-<span class="st"> </span>gtex <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(Name, Bladder, Kidney...Cortex) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb46-2" title="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">rename</span>(<span class="dt">gene =</span> Name, <span class="dt">Kidney =</span> Kidney...Cortex) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb46-3" title="3"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">gene =</span> <span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">..*&quot;</span>,<span class="st">&quot;&quot;</span>,gene)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb46-4" title="4"><span class="st">  </span><span class="kw">pivot_longer</span>(<span class="op">!</span>gene, <span class="dt">names_to=</span><span class="st">&quot;tissue&quot;</span>, <span class="dt">values_to=</span><span class="st">&quot;tpm&quot;</span>)</a></code></pre></div>
<p>Now define two vectors of genes that are specific to bladder and kidney:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" title="1">bladder_expr &lt;-<span class="st"> </span>tissues <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb47-2" title="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(tissue <span class="op">==</span><span class="st"> &quot;Bladder&quot;</span> <span class="op">&amp;</span><span class="st"> </span>tpm <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb47-3" title="3"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">pull</span>(gene)</a>
<a class="sourceLine" id="cb47-4" title="4">kidney_expr &lt;-<span class="st"> </span>tissues <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb47-5" title="5"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(tissue <span class="op">==</span><span class="st"> &quot;Kidney&quot;</span> <span class="op">&amp;</span><span class="st"> </span>tpm <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb47-6" title="6"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">pull</span>(gene)</a>
<a class="sourceLine" id="cb47-7" title="7">int &lt;-<span class="st"> </span><span class="kw">intersect</span>(bladder_expr, kidney_expr)</a>
<a class="sourceLine" id="cb47-8" title="8">bladder_expr &lt;-<span class="st"> </span><span class="kw">setdiff</span>(bladder_expr, int)</a>
<a class="sourceLine" id="cb47-9" title="9">kidney_expr &lt;-<span class="st"> </span><span class="kw">setdiff</span>(kidney_expr, int)</a></code></pre></div>
<p>Next, use an existing TxDb to locate these genes in the genomes. While
we usually recommend to use GENCODE genes for human analysis, because
the ENCODE chromatin modification peak files on AnnotationHub are in
hg19, we use the UCSC hg19 genes here for simplicity of the code:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" title="1"><span class="kw">library</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</a>
<a class="sourceLine" id="cb48-2" title="2"><span class="kw">library</span>(org.Hs.eg.db)</a></code></pre></div>
<p>Add the ENSEMBL ID and pull out the two tissue-specific sets.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" title="1">g &lt;-<span class="st"> </span><span class="kw">genes</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</a></code></pre></div>
<pre><code>##   403 genes were dropped because they have exons located on both strands
##   of the same reference sequence or on more than one reference sequence,
##   so cannot be represented by a single genomic range.
##   Use &#39;single.strand.genes.only=FALSE&#39; to get all the genes in a
##   GRangesList object, or use suppressMessages() to suppress this message.</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" title="1"><span class="kw">library</span>(plyranges)</a>
<a class="sourceLine" id="cb51-2" title="2">g &lt;-<span class="st"> </span>g <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">ensembl =</span> <span class="kw">mapIds</span>(org.Hs.eg.db, gene_id, <span class="st">&quot;ENSEMBL&quot;</span>, <span class="st">&quot;ENTREZID&quot;</span>))</a>
<a class="sourceLine" id="cb51-3" title="3">bladder_g &lt;-<span class="st"> </span>g <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(ensembl <span class="op">%in%</span><span class="st"> </span>bladder_expr)</a>
<a class="sourceLine" id="cb51-4" title="4">kidney_g &lt;-<span class="st"> </span>g <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(ensembl <span class="op">%in%</span><span class="st"> </span>kidney_expr)</a></code></pre></div>
<p>Finally we combine the two sets with <code>bind_ranges</code>, and we change
the feature size from the whole gene extent (the range from the
leftmost exon to rightmost exon), to just the TSS, using <code>anchor_5p</code>
and <code>mutate</code>.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" title="1">tss &lt;-<span class="st"> </span><span class="kw">bind_ranges</span>(<span class="dt">bladder=</span>bladder_g,</a>
<a class="sourceLine" id="cb52-2" title="2">                   <span class="dt">kidney=</span>kidney_g,</a>
<a class="sourceLine" id="cb52-3" title="3">                   <span class="dt">.id=</span><span class="st">&quot;gtissue&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb52-4" title="4"><span class="st">  </span><span class="kw">anchor_5p</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb52-5" title="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">width=</span><span class="dv">1</span>)</a></code></pre></div>
<p>Now we will obtain the H3K27ac peak sets:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" title="1"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb53-2" title="2">ah &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()</a>
<a class="sourceLine" id="cb53-3" title="3"><span class="co"># query(ah, c(&quot;Homo sapiens&quot;, &quot;bladder&quot;, &quot;H3K27ac&quot;, &quot;narrowPeak&quot;))</span></a>
<a class="sourceLine" id="cb53-4" title="4">bladder_pks &lt;-<span class="st"> </span>ah[[<span class="st">&quot;AH44180&quot;</span>]]</a>
<a class="sourceLine" id="cb53-5" title="5"><span class="co"># query(ah, c(&quot;Homo sapiens&quot;, &quot;kidney&quot;, &quot;H3K27ac&quot;, &quot;narrowPeak&quot;))</span></a>
<a class="sourceLine" id="cb53-6" title="6">kidney_pks &lt;-<span class="st"> </span>ah[[<span class="st">&quot;AH43443&quot;</span>]]</a></code></pre></div>
<p>We download these and scale so they have the same 90% quantile of <code>signalValue</code>.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" title="1">ninety &lt;-<span class="st"> </span><span class="cf">function</span>(x) <span class="kw">quantile</span>(x, <span class="fl">.9</span>, <span class="dt">names=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb54-2" title="2">bladder_pks &lt;-<span class="st"> </span>bladder_pks <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb54-3" title="3"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">signal =</span> signalValue <span class="op">/</span><span class="st"> </span><span class="kw">ninety</span>(signalValue))</a>
<a class="sourceLine" id="cb54-4" title="4">kidney_pks &lt;-<span class="st"> </span>kidney_pks <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb54-5" title="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">signal =</span> signalValue <span class="op">/</span><span class="st"> </span><span class="kw">ninety</span>(signalValue))</a></code></pre></div>
<p>Combine the peaks from bladder and kidney, filter to those with &lt; 0.1%
FDR, and center the peak on the summit (the <code>peak</code> column gives the
shift from the left side to the summit).</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" title="1">pks &lt;-<span class="st"> </span><span class="kw">bind_ranges</span>(<span class="dt">bladder=</span>bladder_pks,</a>
<a class="sourceLine" id="cb55-2" title="2">                   <span class="dt">kidney=</span>kidney_pks,</a>
<a class="sourceLine" id="cb55-3" title="3">                   <span class="dt">.id=</span><span class="st">&quot;ptissue&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb55-4" title="4"><span class="st">  </span><span class="kw">filter</span>(qValue <span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>, width <span class="op">&lt;=</span><span class="st"> </span><span class="dv">1000</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb55-5" title="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">start =</span> start <span class="op">+</span><span class="st"> </span>peak) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb55-6" title="6"><span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>peak) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb55-7" title="7"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">width =</span> <span class="dv">1</span>)</a></code></pre></div>
<p>Finally, once we have two tidy range sets, the <em>plyranges</em> analysis is
just a few lines. It appears that tissue-specific peaks are enriched
near the tissue-specific genes for both bladder and kidney.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" title="1">tss <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">join_overlap_left</span>(pks, <span class="dt">maxgap=</span><span class="dv">500</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb56-2" title="2"><span class="st">  </span><span class="kw">group_by</span>(gtissue, ptissue) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb56-3" title="3"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">count =</span> <span class="kw">n</span>(), <span class="dt">meanSignal =</span> <span class="kw">mean</span>(signal))</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 4 columns
##   gtissue     ptissue     count meanSignal
##     &lt;Rle&gt; &lt;character&gt; &lt;integer&gt;  &lt;numeric&gt;
## 1 bladder     bladder      3065   1.182096
## 2 bladder      kidney      2100   0.919537
## 3 bladder          NA       862         NA
## 4  kidney     bladder       131   0.985926
## 5  kidney      kidney       261   0.712519
## 6  kidney          NA       243         NA</code></pre>
<p>Whatâs wrong with this analysis?</p>
<ul>
<li>We didnât figure out the expressed promoter, we just looked at the
left or rightmost isoform (for + or - strand genes, respectively).</li>
</ul>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="overlap-granges-with-plyranges.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bootstrap-overlap.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["tidy-ranges-tutorial.pdf", "tidy-ranges-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
